File: src/matfile/Convective_Flux.m
function [Conv_Flux_x, Conv_Flux_y] = ...
    Convective_Flux(FluxSumOld, ...
                        Ucont_x, Ucont_y, Ucat_x, Ucat_y, ...
                        M, N, Nghost, dx, dy, ...
                        iphys, iphye, jphys, jphye, DEBUG)

    Conv_Flux_x = FluxSumOld.Convective.Flux_x;
    Conv_Flux_y = FluxSumOld.Convective.Flux_y;

    M3 = M+1;
    N3 = N+1;

    G = Nghost;

    coef = 1/8;

    for ii = iphys:iphye
        for jj = jphys:jphye
            Conv_Flux_y(jj, ii) = 0;
            Conv_Flux_x(jj, ii) = 0;
        end
    end

    % All the hafl node fluxes
    %% x - contribution
    if DEBUG
        fprintf('Fpx1 at \n');
    end
    %% ----------------------------------
    for i = 1:M3
        for j = 1:N
            ucon = Ucont_x(j, i) /  2;
            
            up = ucon + abs(ucon); % == 0 if ucont_x(j, i) < 0
            um = ucon - abs(ucon); % == 0 if ucont_x(j, i) > 0

            if DEBUG
                fprintf('\t (%d, %d) has Ucont_x that ', i, j);
            end

            if (i == 1)
                Fpx1(j, i) = um * ( coef * ( - Ucat_x(j+G, i+G+2) - 2 * Ucat_x(j+G, i+G+1) + 3 * Ucat_x(j+G, i+G)  ) + Ucat_x(j+G, i+G+1) ) ...
                           + up * ( coef * ( - Ucat_x(j+G, i+G)   - 2 * Ucat_x(j+G, i+G)   + 3 * Ucat_x(j+G, i+G+1)) + Ucat_x(j+G, i+G)   );
                
                if DEBUG
                    if up
                        fprintf('> 0 and used Ucat_x at (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G+2, j+G, i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            elseif (i == M3)
                Fpx1(j, i) =  um * ( coef * ( - Ucat_x(j+G, i+G+1) - 2 * Ucat_x(j+G, i+G+1) + 3 * Ucat_x(j+G, i+G)  ) + Ucat_x(j+G, i+G+1) ) ...
                            + up * ( coef * ( - Ucat_x(j+G, i+G-1) - 2 * Ucat_x(j+G, i+G)   + 3 * Ucat_x(j+G, i+G+1)) + Ucat_x(j+G, i+G)   );
                        
                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_x at (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            else
                Fpx1(j, i) =  um * ( coef * ( - Ucat_x(j+G, i+G+2) - 2 * Ucat_x(j+G, i+G+1) + 3 * Ucat_x(j+G, i+G)  ) + Ucat_x(j+G, i+G+1) ) ...
                            + up * ( coef * ( - Ucat_x(j+G, i+G-1) - 2 * Ucat_x(j+G, i+G)   + 3 * Ucat_x(j+G, i+G+1)) + Ucat_x(j+G, i+G)   );

                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G+2, j+G, i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            end
        end
    end

    if DEBUG
        fprintf('Fpx2 at \n');
    end
    %% ----------------------------------
    for i = 1:M
        for j = 1:N3
            
            ucon = Ucont_y(j, i) /  2;
            
            up = ucon + abs(ucon);
            um = ucon - abs(ucon);

            if DEBUG
                fprintf('\t (%d, %d) has Ucont_y that ', i, j);
            end
        
            if (j == 1)
                Fpx2(j, i) = um * ( coef * ( - Ucat_x(j+G+2, i+G) - 2 * Ucat_x(j+G+1, i+G) + 3 * Ucat_x(j+G,   i+G) ) + Ucat_x(j+G+1, i+G) ) ...
                           + up * ( coef * ( - Ucat_x(j+G,   i+G) - 2 * Ucat_x(j+G,   i+G) + 3 * Ucat_x(j+G+1, i+G))  + Ucat_x(j+G,   i+G) );
                
                if DEBUG
                    if up
                        fprintf('> 0 and used Ucat_x at (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+2, i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            elseif (j == N3)
                Fpx2(j, i) =  um * ( coef * ( - Ucat_x(j+G+1, i+G) - 2 * Ucat_x(j+G+1, i+G) + 3 * Ucat_x(j+G,   i+G) ) + Ucat_x(j+G+1, i+G) ) ...
                            + up * ( coef * ( - Ucat_x(j+G-1, i+G) - 2 * Ucat_x(j+G,   i+G) + 3 * Ucat_x(j+G+1, i+G))  + Ucat_x(j+G,   i+G) );
                        
                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_x at (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            else
                Fpx2(j, i) = um * ( coef * ( - Ucat_x(j+G+2, i+G) - 2 * Ucat_x(j+G+1, i+G) + 3 * Ucat_x(j+G,   i+G) ) + Ucat_x(j+G+1, i+G) ) ...
                           + up * ( coef * ( - Ucat_x(j+G-1, i+G) - 2 * Ucat_x(j+G,   i+G) + 3 * Ucat_x(j+G+1, i+G))  + Ucat_x(j+G,   i+G) );

                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_x at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+2, i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            end
        end
    end
    
    %% y - contribution
    if DEBUG
        fprintf('Fpy1 at \n');
    end
    %% ----------------------------------
    for i = 1:M3
        for j = 1:N
            ucon = Ucont_x(j, i) /  2;
            
            up = ucon + abs(ucon);
            um = ucon - abs(ucon);

            if DEBUG
                fprintf('\t (%d, %d) has Ucont_x that ', i, j);
            end

            if (i == 1)
                Fpy1(j, i) = um * ( coef * ( - Ucat_y(j+G, i+G+2) - 2 * Ucat_y(j+G, i+G+1) + 3 * Ucat_y(j+G, i+G)  ) + Ucat_y(j+G, i+G+1) ) ...
                           + up * ( coef * ( - Ucat_y(j+G, i+G)   - 2 * Ucat_y(j+G, i+G)   + 3 * Ucat_y(j+G, i+G+1)) + Ucat_y(j+G, i+G)   );
                
                if DEBUG
                    if up
                        fprintf('> 0 and used Ucat_y at (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G+2, j+G, i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            elseif (i == M3)
                Fpy1(j, i) =  um * ( coef * ( - Ucat_y(j+G, i+G+1) - 2 * Ucat_y(j+G, i+G+1) + 3 * Ucat_y(j+G, i+G)  ) + Ucat_y(j+G, i+G+1) ) ...
                            + up * ( coef * ( - Ucat_y(j+G, i+G-1) - 2 * Ucat_y(j+G, i+G)   + 3 * Ucat_y(j+G, i+G+1)) + Ucat_y(j+G, i+G)   );
                        
                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_y at (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            else
                Fpy1(j, i) =  um * ( coef * ( - Ucat_y(j+G, i+G+2) - 2 * Ucat_y(j+G, i+G+1) + 3 * Ucat_y(j+G, i+G)  ) + Ucat_y(j+G, i+G+1) ) ...
                            + up * ( coef * ( - Ucat_y(j+G, i+G-1) - 2 * Ucat_y(j+G, i+G)   + 3 * Ucat_y(j+G, i+G+1)) + Ucat_y(j+G, i+G)   );

                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G+1, j+G, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G+2, j+G, i+G+1, j+G, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            end
        end
    end

    if DEBUG
        fprintf('Fpy2 at \n');
    end
    %% ----------------------------------
    for i = 1:M
        for j = 1:N3
            
            ucon = Ucont_y(j, i) /  2;
            
            up = ucon + abs(ucon);
            um = ucon - abs(ucon);

            if DEBUG
                fprintf('\t (%d, %d) has Ucont_y that ', i, j);
            end
        
            if (j == 1)
                Fpy2(j, i) = um * ( coef * ( - Ucat_y(j+G+2, i+G) - 2 * Ucat_y(j+G+1, i+G) + 3 * Ucat_y(j+G,   i+G) ) + Ucat_y(j+G+1, i+G) ) ...
                           + up * ( coef * ( - Ucat_y(j+G,   i+G) - 2 * Ucat_y(j+G,   i+G) + 3 * Ucat_y(j+G+1, i+G))  + Ucat_y(j+G,   i+G) );
                
                if DEBUG
                    if up
                        fprintf('> 0 and used Ucat_y at (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G);
                    elseif um
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+2, i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            elseif (j == N3)
                Fpy2(j, i) =  um * ( coef * ( - Ucat_y(j+G+1, i+G) - 2 * Ucat_y(j+G+1, i+G) + 3 * Ucat_y(j+G,   i+G) ) + Ucat_y(j+G+1, i+G) ) ...
                            + up * ( coef * ( - Ucat_y(j+G-1, i+G) - 2 * Ucat_y(j+G,   i+G) + 3 * Ucat_y(j+G+1, i+G))  + Ucat_y(j+G,   i+G) );
   
                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_y at (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            else
                Fpy2(j, i) = um * ( coef * ( - Ucat_y(j+G+2, i+G) - 2 * Ucat_y(j+G+1, i+G) + 3 * Ucat_y(j+G,   i+G) ) + Ucat_y(j+G+1, i+G) ) ...
                           + up * ( coef * ( - Ucat_y(j+G-1, i+G) - 2 * Ucat_y(j+G,   i+G) + 3 * Ucat_y(j+G+1, i+G))  + Ucat_y(j+G,   i+G) );

                if DEBUG
                    if up 
                        fprintf('< 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+1, i+G, j+G, i+G-1, j+G);
                    elseif um
                        fprintf('> 0 and used Ucat_y at (%d, %d), (%d, %d), (%d, %d) \n', i+G, j+G+2, i+G, j+G+1, i+G, j+G);
                    else
                        fprintf('= 0 and thus, = 0 \n');
                    end
                end
            end
        end
    end

    % Take the derivatives first term
    for i = iphys:iphye
        for j = jphys:jphye
            Conv_Flux_x(j, i) = (Fpx1(j-G, i-G+1) - Fpx1(j-G, i-G)) / dx  +  (Fpx2(j-G+1, i-G) - Fpx2(j-G, i-G)) / dy; 
            Conv_Flux_y(j, i) = (Fpy1(j-G, i-G+1) - Fpy1(j-G, i-G)) / dx  +  (Fpy2(j-G+1, i-G) - Fpy2(j-G, i-G)) / dy;
        end
    end

end

File: src/jul/momentum/right-hand-side/convective_flux.jl
module ConvectiveFlux
   export rhs_convective_flux_calc

   function rhs_convective_flux_calc(CSolution, coeff)
      let M = deepcopy(CSolution.m2)
          N = deepcopy(CSolution.n2)
          dx = deepcopy(CSolution.dx)
          dy = deepcopy(CSolution.dy)
          ucat_x = deepcopy(CSolution.ucat_x)
          ucat_y = deepcopy(CSolution.ucat_y)
          ucur_x = deepcopy(CSolution.ucur_x)
          ucur_y = deepcopy(CSolution.ucur_y)

          fpx1 = Matrix{Float64}(undef, M, N)
          fpy1 = Matrix{Float64}(undef, M, N)
          fpx2 = Matrix{Float64}(undef, M, N)
          fpy2 = Matrix{Float64}(undef, M, N)

          conv_flux_x = Matrix{Float64}(undef, M, N)
          conv_flux_y = Matrix{Float64}(undef, M, N)

         for ii = 1:M-1
            for jj = 1:N
               un = 0.5*ucur_x[ii, jj]
               um = un - abs(un)
               up = un + abs(un)

               vn = 0.5*ucur_y[ii, jj]
               vm = vn - abs(vn)
               vp = vn + abs(vn)

               if ( ii == 1 )
                  fpx1[ii,  jj] = um*( coeff*( - ucat_x[ii+2, jj] - 2*ucat_x[ii+1, jj] + 3*ucat_x[ii, jj] ) + ucat_x[ii+1, jj] ) + up*( coeff*( - ucat_x[ii, jj] - 2*ucat_x[ii, jj] + 3*ucat_x[ii+1, jj] ) + ucat_x[ii, jj] )
                  
                  fpy1[ii,  jj] = vm*( coeff*( - ucat_y[ii+2, jj] - 2*ucat_y[ii+1, jj] + 3*ucat_y[ii, jj] ) + ucat_y[ii+1, jj] ) + vp*( coeff*( - ucat_y[ii, jj] - 2*ucat_y[ii, jj] + 3*ucat_y[ii+1, jj] ) + ucat_y[ii, jj] )
               elseif ( ii == M-1 )
                  fpx1[ii,  jj] = um*( coeff*( - ucat_x[ii+1, jj] - 2*ucat_x[ii+1, jj] + 3*ucat_x[ii, jj] ) + ucat_x[ii+1, jj] ) + up*( coeff*( - ucat_x[ii-1, jj] - 2*ucat_x[ii, jj] + 3*ucat_x[ii+1, jj] ) + ucat_x[ii, jj] )
                  
                  fpy1[ii,  jj] = vm*( coeff*( - ucat_y[ii+1, jj] - 2*ucat_y[ii+1, jj] + 3*ucat_y[ii, jj] ) + ucat_y[ii+1, jj] ) + vp*( coeff*( - ucat_y[ii-1, jj] - 2*ucat_y[ii, jj] + 3*ucat_y[ii+1, jj] ) + ucat_y[ii, jj] )
               else
                  fpx1[ii,  jj] = um*( coeff*( - ucat_x[ii+2, jj] - 2*ucat_x[ii+1, jj] + 3*ucat_x[ii, jj] ) + ucat_x[ii+1, jj] ) + up*( coeff*( - ucat_x[ii-1, jj] - 2*ucat_x[ii, jj] + 3*ucat_x[ii+1, jj] ) + ucat_x[ii, jj] )
                  
                  fpy1[ii,  jj] = vm*( coeff*( - ucat_y[ii+2, jj] - 2*ucat_y[ii+1, jj] + 3*ucat_y[ii, jj] ) + ucat_y[ii+1, jj] ) + vp*( coeff*( - ucat_y[ii-1, jj] - 2*ucat_y[ii, jj] + 3*ucat_y[ii+1, jj] ) + ucat_y[ii, jj] )
               end
            end
         end

         for jj = 1:N
            fpx1[M, jj] = 0
            fpy1[M, jj] = 0
         end

         for ii = 1:M
            for jj = 1:N-1
               un = 0.5*ucur_x[ii, jj]
               um = un - abs(un)
               up = un + abs(un)

               vn = 0.5*ucur_y[ii, jj]
               vm = vn - abs(vn)
               vp = vn + abs(vn)

               if ( jj == 1 )
                  fpx2[ii,  jj] = um*( coeff*( - ucat_x[ii, jj+2] - 2*ucat_x[ii, jj+1] + 3*ucat_x[ii, jj] ) + ucat_x[ii, jj+1] ) + up*( coeff*( - ucat_x[ii, jj] - 2*ucat_x[ii, jj] + 3*ucat_x[ii, jj+1] ) + ucat_x[ii, jj] )
                  
                  fpy2[ii,  jj] = vm*( coeff*( - ucat_y[ii, jj+2] - 2*ucat_y[ii, jj+1] + 3*ucat_y[ii, jj] ) + ucat_y[ii, jj+1] ) + vp*( coeff*( - ucat_y[ii, jj] - 2*ucat_y[ii, jj] + 3*ucat_y[ii, jj+1] ) + ucat_y[ii, jj] )
               elseif ( jj == N-1 )
                  fpx2[ii,  jj] = um*( coeff*( - ucat_x[ii, jj+1] - 2*ucat_x[ii, jj+1] + 3*ucat_x[ii, jj] ) + ucat_x[ii, jj+1] ) + up*( coeff*( - ucat_x[ii, jj-1] - 2*ucat_x[ii, jj] + 3*ucat_x[ii, jj+1] ) + ucat_x[ii, jj] )
                  
                  fpy2[ii,  jj] = vm*( coeff*( - ucat_y[ii, jj+1] - 2*ucat_y[ii, jj+1] + 3*ucat_y[ii, jj] ) + ucat_y[ii, jj+1] ) + vp*( coeff*( - ucat_y[ii, jj-1] - 2*ucat_y[ii, jj] + 3*ucat_y[ii, jj+1] ) + ucat_y[ii, jj] )
               else
                  fpx2[ii,  jj] = um*( coeff*( - ucat_x[ii, jj+2] - 2*ucat_x[ii, jj+1] + 3*ucat_x[ii, jj] ) + ucat_x[ii, jj+1] ) + up*( coeff*( - ucat_x[ii, jj-1] - 2*ucat_x[ii, jj] + 3*ucat_x[ii, jj+1] ) + ucat_x[ii, jj] )
                  
                  fpy2[ii,  jj] = vm*( coeff*( - ucat_y[ii, jj+2] - 2*ucat_y[ii, jj+1] + 3*ucat_y[ii, jj] ) + ucat_y[ii, jj+1] ) + vp*( coeff*( - ucat_y[ii, jj-1] - 2*ucat_y[ii, jj] + 3*ucat_y[ii, jj+1] ) + ucat_y[ii, jj] )
               end
            end
         end

         for ii = 1:M
            fpx2[ii, N] = 0
            fpy2[ii, N] = 0
         end

         for ii = 2:M-1
            for jj = 2:N-1
               conv_flux_x[ii, jj] = ( fpx1[ii, jj] - fpx1[ii-1, jj] )/dx + ( fpx2[ii, jj] - fpx2[ii, jj-1] )/dy
               conv_flux_y[ii, jj] = ( fpy1[ii, jj] - fpy1[ii-1, jj] )/dx + ( fpy2[ii, jj] - fpy2[ii, jj-1] )/dy
            end
         end

         for ii = 1:M
            conv_flux_x[ii, 1] = 0.0
            conv_flux_y[ii, 1] = 0.0
            conv_flux_x[ii, N] = 0.0
            conv_flux_y[ii, N] = 0.0
         end

         for jj = 1:N
            conv_flux_x[1, jj] = 0.0
            conv_flux_y[1, jj] = 0.0
            conv_flux_x[M, jj] = 0.0
            conv_flux_y[M, jj] = 0.0
         end

         return conv_flux_x, conv_flux_y
      end
   end
end